metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -0,0 +1,660 @@
+        self.get_access_token()
+        logger.info("Connected to {}".format(self._get_authority_url()))
+        self._request_session = requests.Session()
+        # set re-try parameter for request_session
+        self._request_session.mount(
+            "https://",
+            HTTPAdapter(
+                max_retries=Retry(
+                    total=3,
+                    backoff_factor=1,
+                    allowed_methods=None,
+                    status_forcelist=[429, 500, 502, 503, 504],
+                )
+            ),
+        )
+
+    @abstractmethod
+    def get_groups_endpoint(self) -> str:
+        pass
+
+    @abstractmethod
+    def get_dashboards_endpoint(self, workspace: Workspace) -> str:
+        pass
+
+    @abstractmethod
+    def get_reports_endpoint(self, workspace: Workspace) -> str:
+        pass
+
+    @abstractmethod
+    def get_tiles_endpoint(self, workspace: Workspace, dashboard_id: str) -> str:
+        pass
+
+    @abstractmethod
+    def _get_pages_by_report(self, workspace: Workspace, report_id: str) -> List[Page]:
+        pass
+
+    @abstractmethod
+    def get_dataset(
+        self, workspace_id: str, dataset_id: str
+    ) -> Optional[PowerBIDataset]:
+        pass
+
+    @abstractmethod
+    def get_users(self, workspace_id: str, entity: str, entity_id: str) -> List[User]:
+        pass
+
+    def _get_authority_url(self):
+        return "{}{}".format(DataResolverBase.AUTHORITY, self.__tenant_id)
+
+    def get_access_token(self):
+        if self.__access_token != "":
+            logger.debug("Returning the cached access token")
+            return self.__access_token
+
+        logger.info("Generating PowerBi access token")
+
+        auth_response = self.__msal_client.acquire_token_for_client(
+            scopes=[DataResolverBase.SCOPE]
+        )
+
+        if not auth_response.get("access_token"):
+            logger.warning(
+                "Failed to generate the PowerBi access token. Please check input configuration"
+            )
+            raise ConfigurationError(
+                "Powerbi authorization failed . Please check your input configuration."
+            )
+
+        logger.info("Generated PowerBi access token")
+
+        self.__access_token = "Bearer {}".format(auth_response.get("access_token"))
+
+        logger.debug(f"{Constant.PBIAccessToken}={self.__access_token}")
+
+        return self.__access_token
+
+    def get_dashboards(self, workspace: Workspace) -> List[Dashboard]:
+        """
+        Get the list of dashboard from PowerBi for the given workspace identifier
+
+        TODO: Pagination. As per REST API doc (https://docs.microsoft.com/en-us/rest/api/power-bi/dashboards/get
+        -dashboards), there is no information available on pagination
+        """
+        dashboard_list_endpoint: str = self.get_dashboards_endpoint(workspace)
+
+        # Hit PowerBi
+        logger.info(f"Request to URL={dashboard_list_endpoint}")
```

Comment:
Not a particularly good log line. Can we clean it up a bit?

Reply:
debug?

Reply:
done